src/Entity/SmsMessageResult.php

```diff
@@ -0,0 +1,224 @@
+        return $report;
+      }
+    }
+    return NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getReports() {
+    $reports = [];
+    foreach ($this->get('reports') as $report) {
+      $reports[] = $report->entity;
+    }
+    return $reports;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setReports(array $reports) {
+    $this->reports->filter(function ($item) { return false; });
+    foreach ($reports as $report) {
+      $this->addReport($report);
+    }
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function addReport(PlainDeliveryReportInterface $report) {
+    $this->reports->appendItem(SmsDeliveryReport::convertFromDeliveryReport($report));
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCreditsBalance() {
+    return $this->get('credits_balance')->value;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setCreditsBalance($balance) {
+    if (is_numeric($balance) || is_null($balance)) {
+      $this->set('credits_balance', $balance);
+    }
+    else {
+      throw new SmsException(sprintf('Credit balance set is a %s', gettype($balance)));
+    }
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCreditsUsed() {
+    return $this->get('credits_used')->value;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setCreditsUsed($credits_used) {
+    if (is_numeric($credits_used) || is_null($credits_used)) {
+      $this->set('credits_used', $credits_used);
+    }
+    else {
+      throw new SmsException(sprintf('Credit used is a %s', gettype($credits_used)));
+    }
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function baseFieldDefinitions(EntityTypeInterface $entity_type) {
+    $fields = parent::baseFieldDefinitions($entity_type);
+
+    $fields['error'] = BaseFieldDefinition::create('string')
+      ->setLabel(t('Error'))
+      ->setDescription(t('The normalized error code, either account_error, flooded, no_credit, invalid_sender, authentication or parameters.'))
+      ->setReadOnly(TRUE)
+      ->setRequired(FALSE);
+
+    $fields['error_message'] = BaseFieldDefinition::create('string')
+      ->setLabel(t('Error message'))
+      ->setDescription(t('The description of the error from the gateway.'))
+      ->setReadOnly(TRUE)
+      ->setDefaultValue('')
+      ->setRequired(FALSE);
+
+    $fields['credits_balance'] = BaseFieldDefinition::create('decimal')
+      ->setLabel(t('Credits balance'))
+      ->setDescription(t('The balance of credits after the message was sent.'))
+      ->setReadOnly(TRUE)
+      ->setRequired(TRUE);
+
+    $fields['credits_used'] = BaseFieldDefinition::create('decimal')
+      ->setLabel(t('Credits used'))
+      ->setDescription(t('The credits used for the message transaction.'))
+      ->setReadOnly(TRUE)
+      ->setRequired(TRUE);
+
+    $fields['reports'] = BaseFieldDefinition::create('entity_reference')
```

Comment:
*As a parent relationship to a message entity. See **relationships** in https://www.drupal.org/node/2836157#comment-11930789